### PR TITLE
Use copy for some operations instead of hardlinks

### PR DESCRIPTION
**Pull request by alesapin**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by tavplubix** [comment]
Sorry, it was misclick

**Review by tavplubix — APPROVED** [review]

File: src/Storages/MergeTree/MutateTask.cpp [hunk 1]
```diff
@@ -1621,7 +1627,13 @@ bool MutateTask::prepare()
             LOG_TRACE(ctx->log, "Part {} doesn't change up to mutation version {} (optimized)", ctx->source_part->name, ctx->future_part->part_info.mutation);
             /// new_data_part is not used here, another part is created instead (see the comment above)
             ctx->temporary_directory_lock = {};
-            auto [part, lock] = ctx->data->cloneAndLoadDataPartOnSameDisk(ctx->source_part, "tmp_mut_", ctx->future_part->part_info, ctx->metadata_snapshot, ctx->txn, &ctx->hardlinked_files, false);
+            NameSet files_to_copy_instead_of_hardlinks;
+            auto settings_ptr = ctx->data->getSettings();
+            bool copy_checksumns = ctx->data->supportsReplication() && settings_ptr->allow_remote_fs_zero_copy_replication && ctx->source_part->isStoredOnRemoteDiskWithZeroCopySupport();
```
Comment (alexey-milovidov) [inline src/Storages/MergeTree/MutateTask.cpp:1632]: Please add a comment explaining "why".

File: src/Storages/MergeTree/MutateTask.cpp [hunk 2]
```diff
@@ -1507,8 +1507,14 @@ bool MutateTask::prepare()
     if (ctx->source_part->isStoredOnDisk() && !isStorageTouchedByMutations(
         ctx->storage_from_source_part, ctx->metadata_snapshot, ctx->commands_for_part, Context::createCopy(context_for_reading)))
     {
+        NameSet files_to_copy_instead_of_hardlinks;
+        auto settings_ptr = ctx->data->getSettings();
+        bool copy_checksumns = ctx->data->supportsReplication() && settings_ptr->allow_remote_fs_zero_copy_replication && ctx->source_part->isStoredOnRemoteDiskWithZeroCopySupport();
+        if (copy_checksumns)
```
Comment (alexey-milovidov) [inline src/Storages/MergeTree/MutateTask.cpp:1513]: Please add a comment explaining "why".

File: src/Storages/MergeTree/IMergeTreeDataPart.cpp [hunk 3]
```diff
@@ -1516,12 +1516,15 @@ void IMergeTreeDataPart::renameToDetached(const String & prefix, DataPartStorage
 
 void IMergeTreeDataPart::makeCloneInDetached(const String & prefix, const StorageMetadataPtr & /*metadata_snapshot*/) const
 {
+    auto storage_settings = storage.getSettings();
+    bool copy_instead_of_hardlink = isStoredOnRemoteDiskWithZeroCopySupport() && storage.supportsReplication() && storage_settings->allow_remote_fs_zero_copy_replication;
```
Comment (alexey-milovidov) [inline src/Storages/MergeTree/IMergeTreeDataPart.cpp:1520]: Please add a comment explaining "why".

File: src/Storages/MergeTree/IDataPartStorage.h [hunk 4]
```diff
@@ -197,7 +197,8 @@ class IDataPartStorage
         const std::string & dir_path,
         bool make_source_readonly,
         std::function<void(const DiskPtr &)> save_metadata_callback,
-        bool copy_instead_of_hardlink) const = 0;
+        bool copy_instead_of_hardlink,
```
Comment (alexey-milovidov) [inline src/Storages/MergeTree/IDataPartStorage.h:200]: Please also make more detailed comments for this method.

**Review by alexey-milovidov — COMMENTED** [review]
If something was incorrect, can we make a test for it?

**Comment by alesapin** [comment]
> If something was incorrect, can we make a test for it?

Yes, actually we can, but it will be quite complex and very flaky....

**Comment by alesapin** [comment]
https://s3.amazonaws.com/clickhouse-test-reports/41832/905d4585acf126aac279e5ca40f4a6fe03daaf74/integration_tests_flaky_check__asan_.html let's merge

**Comment by tavplubix** [comment]
Stress test (msan) - `<Error> test_xmxown6r.concurrent_mutate_mt_4: while loading part all_20_20_0_31 on path data/test_xmxown6r/concurrent_mutate_mt_4/all_20_20_0_31: Code: 499. DB::Exception: The specified key does not exist.: while reading key: test/rbn/dejyhbdyewrsiioulexxmaqruoeno, from bucket: test. (S3_ERROR)`

**Comment by tavplubix** [comment]
However, it happened in BC check, so probably it's already fixed in master

**Pull request merged**